File: 2021/Day 17 - Trick Shot/Part2.cpp
#include <iostream>
#include <math.h>
using namespace std;

int main(){
    int lx, rx, ly, uy;
    scanf("target area: x=%d..%d, y=%d..%d", &lx, &rx, &ly, &uy);
    printf("target area: x=%d..%d, y=%d..%d\n", lx, rx, ly, uy);
    int ans = ((rx - lx) + 1) * ((abs(ly) - abs(uy)) + 1);
    int vy, vx = 3;
    while(true){
        int r = vx * (vx + 1) / 2;
        if(r >= lx && r <= rx){
            break;
        }
        vx++;
    }
    
    vy = 100;
    bool found = false;
    while(true){
        int r = vy * (vy + 1) / 2;
        int s = -(vy + 1);
        while(true){
            if(s >= ly && s <= uy){
                found = true;
                break;
            }
            if(s < ly){
                break;
            }
            s += (s - 1);
        }
        if(found) {
            cout << r << endl;;
            break;
        }
        vy--;
    }
    for(int i = vx; i <= rx / 2; i++){
        for(int j = uy + 1; j <= vy; j++){
            auto [sx, sy] = pair<int,int>(0,0);
            auto dx = i, dy = j;
            while(true){
                if(sx >= lx && sx <= rx && sy >= ly && sy <= uy){
                    cout << i << " " << j << endl;
                    ans++;
                    break;
                }

                if(sx > rx || sy < ly) break;
                sx += dx;
                sy += dy;
                dx--;
                if(dx < 0) dx = 0;
                dy--;
            }
        }
    }
    cout << ans;
    return 0;
}

File: 2023/Day 11 - Cosmic Expansion/Part2.kt
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.max

fun main() {
	val input = generateSequence(::readLine).toList()
	val multiplier = 1000000
	val rowSpace = mutableListOf<Int>()
	val colSpace = mutableListOf<Int>()
	input.forEachIndexed { i, line ->
		for (j in line.indices) {
			if (input[i][j] != '.') {
				return@forEachIndexed
			}
		}
		rowSpace.add(i)
	}

	for(j in input[0].indices) {
		var cleanCol = true
		for(i in input.indices) {
			if (input[i][j] != '.') {
				cleanCol = false
				break
			}
		}
		if (!cleanCol) continue
		colSpace.add(j)
	}

	val galaxies = mutableListOf<Pair<Int, Int>>()
	for (i in input.indices) {
		for (j in input[i].indices) {
			if (input[i][j] == '#') {
				galaxies.add(Pair(i,j))
			}
		}
	}
	var ans: Long = 0L
	for (i in galaxies.indices) {
		for (j in (i+1) until galaxies.size) {
			val myRowRange = min(galaxies[i].first, galaxies[j].first)..max(galaxies[i].first, galaxies[j].first)
			var rowOccurrences = 0
			rowSpace.forEach { 
				if (myRowRange.contains(it)) {
					rowOccurrences++
				}
			}
			val myColRange = min(galaxies[i].second, galaxies[j].second)..max(galaxies[i].second, galaxies[j].second)
			var colOccurrences = 0
			colSpace.forEach { 
				if (myColRange.contains(it)) {
					colOccurrences++
				}
			}
			ans += abs(galaxies[i].first - galaxies[j].first) + (rowOccurrences * (multiplier-1)) + abs(galaxies[i].second - galaxies[j].second) + (colOccurrences * (multiplier-1))
		}
	}
	println(ans)
}